xclim/indices/_agro.py

```diff
@@ -589,6 +595,225 @@ def water_budget(
+)
+def standardized_precipitation_index(
+    pr: xarray.DataArray,
+    pr_cal: xarray.DataArray,
+    freq: str = "MS",
+    window: int = 1,
+    dist: str = "gamma",
+    method: str = "ML",
+) -> xarray.DataArray:
+    r"""Standardized Precipitation Index (SPI).
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+      Daily precipitation.
+    pr_cal : xarray.DataArray
+      Daily precipitation used for calibration. Usually this is a temporal subset of `pr` over some reference period.
+    freq : str
+      Resampling frequency. A monthly or daily frequency is expected.
+    window : int
+      Averaging window length relative to the resampling frequency. For example, if `freq="MS"`, i.e. a monthly resampling, the window
+      is an integer number of months.
+    dist : str
+      Name of the univariate distribution, such as `beta`, `expon`, `genextreme`, `gamma`, `gumbel_r`, `lognorm`, `norm`
+      (see :py:mod:`scipy.stats`).
+    method : str
+      Name of the fitting method, such as `ML` (maximum likelihood), `APP` (approximate). The approximate method
+      simply uses the parameters output of `_fit_start`.
+
+    Returns
+    -------
+    xarray.DataArray,
+      Standardized Precipitation Index.
+
+
+    Notes
+    -----
+    The length `N` of the N-month SPI is determined by choosing the `window = N`. Supported statistical distributions are: ["gamma"]
+
+
+    Example
+    -------
+    Computing SPI-3 months using a gamma distribution for the fit
+
+    .. code-block:: python
+
+        import xclim.indices as xci
+        import xarray as xr
+
+        ds = xr.open_dataset(filename)
+        pr = ds.pr
+        pr_cal = pr.sel(time=slice(calibration_start_date, calibration_end_date))
+        spi_3 = xci.standardized_precipitation_index(
+            pr, pr_cal, freq="MS", window=3, dist="gamma", method="ML"
+        )
+
+
+    References
+    ----------
+    McKee, Thomas B., Nolan J. Doesken, and John Kleist. "The relationship of drought frequency and duration to time scales." In Proceedings of the 8th Conference on Applied Climatology, vol. 17, no. 22, pp. 179-183. 1993.
+    """
+    # "WPM" method doesn't seem to work for gamma or pearson3
+    dist_and_methods = {"gamma": ["ML", "APP"]}
+    if dist not in dist_and_methods.keys():
+        raise NotImplementedError(f"The distribution `{dist}` is not supported.")
+    elif method not in dist_and_methods[dist]:
+        raise NotImplementedError(
+            f"The method `{method}` is not supported for distribution `{dist}`."
+        )
+
+    # Determine group type
+    if freq == "D" or freq is None:
+        freq = "D"
+        group = "time.dayofyear"
+    else:
+        _, base, _, _ = parse_offset(freq)
+        if base in ["Y", "A", "Q", "M"]:
+            group = "time.month"
+        else:
+            raise NotImplementedError(f"Resampling frequency `{freq}` not supported.")
+
+    # Resampling precipitations
+    if freq != "D":
```

Comment:
Question of someone who is maybe confused : Does it even make sense to calculate SPEI on ['Y', 'A', 'Q'] ?   Or should we only support daily and monthly frequency?

Reply:
I have seen some definitions of SPI as "annual-SPI-1" which is distinguished from "SPI-12". See for instance [1]. 
* "SPI-12" is resampled over month (`freq=MS`), with a window of 12 months; 
* "annual-SPI-1" is resampled over year (`freq=YS`) with a window of 1 year.

This definition doesn't seem to be used in many research articles though. The expression "annual SPI" is sometimes used as a synonym for "SPI-12" too.  The `monocongo` library doesn't support this definition of SPI either. 

All in all, including this possibility doesn't cost much, but if you feel it could bring confusion, we could decide to forbid this.

[1] Aryal, 2022: https://mdpi-res.com/d_attachment/earth/earth-03-00025/article_deploy/earth-03-00025-v2.pdf

Reply:
If I understand correctly, SPI-12 on january is the same as annual-SPI-1 ? 
If so, 'Y', and 'Q' are just options to drop the overlapped members of the result, we could _not_ implement them and show in the doc how to retrieve them from the monthly version?
Ex to be clear:
```
spiM12 = spi(pr, pr_cal, freq='MS', window=12)
spiY1 = spiM12.sel(time=(spiM12.time.dt.month == 1))
```

Reply:
Right, what you describe is pretty much the same I agree. The only difference I see: The annual-SPI-1 is like having the Januaries of SPI-12 *with a window rolling forward in time* if `freq=YS` is used.